esp/services/ws_packageprocess/ws_packageprocessService.cpp

```diff
@@ -382,11 +414,46 @@ class PackageMapUpdater
         }
         makePackageActive(pkgSet, psEntry, target, checkFlag(PKGADD_MAP_ACTIVATE));
     }
+    void doCreate(const char *partname, const char *xml, unsigned updateFlags, StringArray &filesNotFound)
+    {
+        Owned<IPropertyTree> pTree = createPTreeFromXMLString(xml, ipt_ordered);
+        doCreate(partname, pTree, updateFlags, filesNotFound);
+    }
     void create(const char *partname, const char *xml, unsigned updateFlags, StringArray &filesNotFound)
     {
         init();
         doCreate(partname, xml, updateFlags, filesNotFound);
     }
+    void copy(IPropertyTree *pm, const char *name, unsigned updateFlags, StringArray &filesNotFound)
+    {
+        init();
+        doCreate(name, pm, updateFlags, filesNotFound);
+    }
+    void copy(const char *srcAddress, const char *srcTarget, const char *name, unsigned updateFlags, StringArray &filesNotFound)
+    {
+        VStringBuffer url("http://%s/WsPackageProcess", (srcAddress && *srcAddress) ? srcAddress : ".:8010");
+        Owned<IClientWsPackageProcess> client = createWsPackageProcessClient();
+        StringBuffer user, pw;
+        if (userdesc)
+            client->setUsernameToken(userdesc->getUserName(user), userdesc->getPassword(pw), nullptr);
+        client->addServiceUrl(url);
+
+        StringBuffer scopedPMID;
+        if (srcTarget && *srcTarget)
+            scopedPMID.append(srcTarget).append("::");
```

Comment:
I think we've talked about this before and I don't remember what we concluded regarding multiple append calls versus a single appendf, but I would consider opting for scopedPMID.setf("%s::" srcTarget); here